modules/ml/src/main/java/org/apache/ignite/ml/optimization/GradientDescent.java

```diff
@@ -0,0 +1,194 @@
+     * @return This gradient descent instance
+     */
+    public GradientDescent withConvergenceTol(double convergenceTol) {
+        if (convergenceTol < 0)
+            throw new IllegalArgumentException("Convergence tolerance must be non-negative but got " + convergenceTol);
+        this.convergenceTol = convergenceTol;
+        return this;
+    }
+
+    /**
+     * Computes point where loss function takes minimal value.
+     *
+     * @param data Inputs parameters of loss function
+     * @param initialWeights Initial weights
+     * @return Point where loss function takes minimal value
+     */
+    public Vector optimize(Matrix data, Vector initialWeights) {
+        Vector weights = initialWeights, oldWeights = null, oldGradient = null;
+        if (data instanceof SparseDistributedMatrix) {
+            for (int iteration = 0; iteration < maxIterations; iteration++) {
+                Vector gradient = calculateDistributedGradient((SparseDistributedMatrix)data, weights);
+                Vector newWeights = updater.compute(oldWeights, oldGradient, weights, gradient, iteration);
+                if (isConverged(weights, newWeights))
+                    return newWeights;
+                else {
+                    oldGradient = gradient;
+                    oldWeights = weights;
+                    weights = newWeights;
+                }
+            }
+        }
+        else {
+            Matrix inputs = extractInputs(data);
+            Vector groundTruth = extractGroundTruth(data);
+            for (int iteration = 0; iteration < maxIterations; iteration++) {
+                Vector gradient = lossGradient.compute(inputs, groundTruth, weights);
+                Vector newWeights = updater.compute(oldWeights, oldGradient, weights, gradient, iteration);
+                if (isConverged(weights, newWeights))
+                    return newWeights;
+                else {
+                    oldGradient = gradient;
+                    oldWeights = weights;
+                    weights = newWeights;
+                }
+            }
+        }
+        return weights;
+    }
+
+    /**
+     * Calculates gradient based in distributed matrix using {@link SparseDistributedMatrixMapReducer}.
+     *
+     * @param data Distributed matrix
+     * @param weights Point to calculate gradient
+     * @return Gradient
+     */
+    private Vector calculateDistributedGradient(SparseDistributedMatrix data, Vector weights) {
+        SparseDistributedMatrixMapReducer mapReducer = new SparseDistributedMatrixMapReducer(data);
+        return mapReducer.mapReduce(
+            (matrix, args) -> {
+                Matrix inputs = extractInputs(matrix);
+                Vector groundTruth = extractGroundTruth(matrix);
+                return lossGradient.compute(inputs, groundTruth, args);
+            },
+            gradients -> {
+                Vector resultGradient = new DenseLocalOnHeapVector(data.columnSize());
```

Comment:
What if it is crashed on Local Heap?

Reply:
It's just a resulting vector, it doesn't contain more data than we need. From the other hand it has cardinality equals to cardinality of the feature vector, so it shouldn't be big.